datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMap.java

```diff
@@ -0,0 +1,243 @@
+
+  @Override
+  public void init(DataMapModel dataMapModel) throws MemoryException, IOException {
+    Path indexPath = FileFactory.getPath(dataMapModel.getFilePath());
+    FileSystem fs = FileFactory.getFileSystem(indexPath);
+    if (!fs.exists(indexPath)) {
+      throw new IOException(
+          String.format("Path %s for Bloom index dataMap does not exist", indexPath));
+    }
+    if (!fs.isDirectory(indexPath)) {
+      throw new IOException(
+          String.format("Path %s for Bloom index dataMap must be a directory", indexPath));
+    }
+
+    FileStatus[] indexFileStatus = fs.listStatus(indexPath, new PathFilter() {
+      @Override public boolean accept(Path path) {
+        return path.getName().endsWith(".bloomindex");
+      }
+    });
+    indexFilePath = new String[indexFileStatus.length];
+    indexedColumn = new HashSet<String>();
+    bloomIndexList = new ArrayList<BloomDMModel>();
+    indexCol2BloomDMList = ArrayListMultimap.create();
+    for (int i = 0; i < indexFileStatus.length; i++) {
+      indexFilePath[i] = indexFileStatus[i].getPath().toString();
+      String indexCol = StringUtils.substringBetween(indexFilePath[i], ".carbondata.",
+          ".bloomindex");
+      indexedColumn.add(indexCol);
+      bloomIndexList.addAll(readBloomIndex(indexFilePath[i]));
+      indexCol2BloomDMList.put(indexCol, readBloomIndex(indexFilePath[i]));
+    }
+    LOGGER.info("find bloom index datamap for column: "
+        + StringUtils.join(indexedColumn, ", "));
+  }
+
+  private List<BloomDMModel> readBloomIndex(String indexFile) throws IOException {
+    LOGGER.info("read bloom index from file: " + indexFile);
+    List<BloomDMModel> bloomDMModelList = new ArrayList<BloomDMModel>();
+    DataInputStream dataInStream = null;
+    ObjectInputStream objectInStream = null;
+    try {
+      dataInStream = FileFactory.getDataInputStream(indexFile, FileFactory.getFileType(indexFile));
+      objectInStream = new ObjectInputStream(dataInStream);
+      try {
+        BloomDMModel model = null;
+        while ((model = (BloomDMModel) objectInStream.readObject()) != null) {
+          LOGGER.info("read bloom index: " + model);
+          bloomDMModelList.add(model);
+        }
+      } catch (EOFException e) {
+        LOGGER.info("read " + bloomDMModelList.size() + " bloom indices from " + indexFile);
+      }
+      return bloomDMModelList;
+    } catch (ClassNotFoundException e) {
+      LOGGER.error("Error occrus while reading bloom index");
+      throw new RuntimeException("Error occrus while reading bloom index", e);
+    } finally {
+      CarbonUtil.closeStreams(objectInStream, dataInStream);
+    }
+  }
+
+  @Override
+  public List<Blocklet> prune(FilterResolverIntf filterExp, SegmentProperties segmentProperties,
+      List<PartitionSpec> partitions) throws IOException {
+    List<Blocklet> hitBlocklets = new ArrayList<Blocklet>();
+    if (filterExp == null) {
+      return null;
```

Comment:
better to return an empty list

Reply:
I learned it LuceneDataMap.

Besides, null and empty is not the same here. Empty means that we have pruned all the blocklets here, meaning there is no blocklet to scan later.

Reply:
Oh, then we should document it clearly somewhere, otherwise it is confusing

Reply:
:ok_hand:  fine~